src/operator/tensor/elemwise_unary_op.h

```diff
@@ -31,173 +31,381 @@
 #include "../mshadow_op.h"
 #include "../elemwise_op_common.h"
 #include "../special_functions-inl.h"
-#include "./broadcast_reduce-inl.h"
-#include "./init_op.h"
+#include "../mxnet_op.h"
+#include "cast_storage-inl.h"
 
 namespace mxnet {
 namespace op {
-template<typename xpu, typename op>
-void UnaryLaunch(const nnvm::NodeAttrs& attrs,
-                 const OpContext& ctx,
-                 const std::vector<TBlob>& inputs,
-                 const std::vector<OpReqType>& req,
-                 const std::vector<TBlob>& outputs) {
-  using namespace mshadow;
-  using namespace mxnet_op;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
 
-  CHECK_EQ(inputs.size(), 1U);
-  CHECK_EQ(outputs.size(), 1U);
-  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
-    Kernel<op, xpu>::Launch(s, outputs[0].Size(),
-      outputs[0].dptr<DType>(), inputs[0].dptr<DType>());
-  });
-}
+class OpBase {
+ public:
+  template<int Req>
+  struct MapSetToScalar {
+    template<typename DType>
+    MSHADOW_XINLINE static void Map(int i, DType *out, const DType value) {
+      KERNEL_ASSIGN(out[i], Req, value);
+    }
+  };
 
-template<typename GRAD_OP>
-struct unary_bwd {
-  template<typename DType>
-  MSHADOW_XINLINE static DType Map(DType a, DType b) {
-    return DType(a*GRAD_OP::Map(b));
+ protected:
+  /*! \brief Copy blob data */
+  template<typename xpu>
+  static void inline CopyBlob(mshadow::Stream<xpu> *s,
```

Comment:
Since it's just copying data from one tblob to another, do we really need a function as a wrapper of mshadow::Copy? In addition, this function has a general name but only supports copying data between the same type of contexts (`xpu`). Can we just use mshadow::Copy?